doc/PROTOCOL-SEMANTICS.md

```diff
@@ -0,0 +1,268 @@
+
+Metadata has keys with associated values. Each key can have multiple values.
+Keys are unordered, but values for a key are ordered. Keys are case insensitive,
+and commonly canonicalized by making lower case. APIs are permitted to require a
+canonical representation and that representation may be different than the
+"lower case" representation mentioned here. A key can be for ASCII or binary
+values. If a key is for binary values, its name must be suffixed with "-bin".
+Otherwise it is for ASCII values.
+
+ASCII values are discouraged from having leading or trailing whitespace. If such
+a value contains leading or trailing whitespace, the whitespace may be stripped.
+Multiple ASCII values for the same key may be joined together with "," (a comma)
+as the delimiter and be considered semantically equivalent to the multi-value
+form.  However, such a transformation is lossy; an arbitrary ASCII value may not
+be split on comma and be assumed to be equivalent to a valid multi-value form
+for its key.
+
+The position of the Metadata in either Headers or Trailers is semantically
+important. Metadata from a Headers may not be moved to Trailers or vise-versa
+without additional knowledge of the individual key semantics.
+
+## Additional Features
+
+Although these are "additional" features, that does not make them unimportant.
+The are additional because they do not need to be represented as a fundamental
+gRPC protocol concept.
+
+### Deadline Propagation
+
+It is a common concept for RPCs to have a Deadline, a point in time by which the
+Call needs to complete before the client will give up. If the Client has a
+Deadline for the Call it should be included in the Request Headers. Calls are
+not required to have a Deadline. If a Call has a Deadline, both the client and
+server should track the Deadline and Cancel the Call when the Deadline is
+reached. When the client Cancels the Call, it should report a Trailers
+containing a Status Code of DEADLINE_EXCEEDED.
+
+While the feature is called "Deadline" and involves Deadlines from the user's
+perspective, it is generally communicated with a timeout, which is a duration
+instead of a point in time. Communicating a duration avoids the need for civil
+time clock synchronization between the client and server.
+
+By its very nature, the client and server's Deadline will be slightly different
+and their detection of its expiration will race. When clients receive a
+Cancellation on a Call with a Deadline, they should double-check whether the
+Deadline has passed. If it has passed, they should override the cancellation
+with Trailers with one that has a Status Code of DEADLINE_EXCEEDED.
+
+### Authority
+
+Virtual hosting is a common concept where a server supports multiple identities
+and supports different services depending on the identity. "Identity" is a
+server name, in "host" or "host:port" form as found for the "authority" portion
+of URIs. Transports able to support virtual hosting should transmit the
```

Comment:
> Transports able to support virtual hosting should transmit the
> authority the client identifies the service as and make that available

I had a hard time parsing this.  Maybe something like?

Transports that support virtual hosting should transmit the authority, as identified by the client, and make it available to the server application.